x/merkledb/benchmarks/benchmark.go

```diff
@@ -0,0 +1,372 @@
+	"github.com/spf13/pflag"
+
+	"github.com/ava-labs/avalanchego/database/leveldb"
+	"github.com/ava-labs/avalanchego/trace"
+	"github.com/ava-labs/avalanchego/utils/logging"
+	"github.com/ava-labs/avalanchego/utils/units"
+	"github.com/ava-labs/avalanchego/x/merkledb"
+)
+
+const (
+	defaultDatabaseEntries    = 2000000
+	databaseCreationBatchSize = 1000000
+	databaseRunningBatchSize  = 2500
+	databaseRunningUpdateSize = 5000
+	defaultMetricsPort        = 3000
+)
+
+var (
+	databaseEntries = pflag.Uint64("n", defaultDatabaseEntries, "number of database entries")
+	httpMetricPort  = pflag.Uint64("p", defaultMetricsPort, "default metrics port")
+	verbose         = pflag.Bool("v", false, "verbose")
+)
+
+func getMerkleDBConfig(promRegistry prometheus.Registerer) merkledb.Config {
+	const defaultHistoryLength = 300
+	return merkledb.Config{
+		BranchFactor:                merkledb.BranchFactor16,
+		Hasher:                      merkledb.DefaultHasher,
+		RootGenConcurrency:          0,
+		HistoryLength:               defaultHistoryLength,
+		ValueNodeCacheSize:          units.MiB,
+		IntermediateNodeCacheSize:   units.MiB,
+		IntermediateWriteBufferSize: units.KiB,
+		IntermediateWriteBatchSize:  256 * units.KiB,
+		Reg:                         promRegistry,
+		TraceLevel:                  merkledb.NoTrace,
+		Tracer:                      trace.Noop,
+	}
+}
+
+func getGoldenStagingDatabaseDirectory() string {
+	wd, _ := os.Getwd()
+	return path.Join(wd, fmt.Sprintf("db-bench-test-golden-staging-%d", *databaseEntries))
+}
+
+func getGoldenDatabaseDirectory() string {
+	wd, _ := os.Getwd()
+	return path.Join(wd, fmt.Sprintf("db-bench-test-golden-%d", *databaseEntries))
+}
+
+func getRunningDatabaseDirectory() string {
+	wd, _ := os.Getwd()
+	return path.Join(wd, fmt.Sprintf("db-bench-test-running-%d", *databaseEntries))
+}
+
+func calculateIndexEncoding(idx uint64) []byte {
+	var entryEncoding [8]byte
+	binary.NativeEndian.PutUint64(entryEncoding[:], idx)
+	entryHash := sha256.Sum256(entryEncoding[:])
+	return entryHash[:]
+}
+
+func createGoldenDatabase() error {
+	err := os.RemoveAll(getGoldenStagingDatabaseDirectory())
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "unable to remove running directory : %v\n", err)
+		return err
+	}
+	err = os.Mkdir(getGoldenStagingDatabaseDirectory(), 0o777)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "unable to create golden staging directory : %v\n", err)
+		return err
+	}
+
+	promRegistry := prometheus.NewRegistry()
+
+	levelDB, err := leveldb.New(
+		getGoldenStagingDatabaseDirectory(),
+		nil,
+		logging.NoLog{},
+		promRegistry,
+	)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "unable to create levelDB database : %v\n", err)
+		return err
+	}
+
+	mdb, err := merkledb.New(context.Background(), levelDB, getMerkleDBConfig(nil))
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "unable to create merkleDB database : %v\n", err)
+		return err
+	}
+
+	fmt.Printf("Initializing database.")
+	ticksCh := make(chan interface{})
```

Comment:
nit: do you really need this ticker? Might be easier to report every 100k rows or something.

Reply:
reporting every 100k rows doesn't work nicely because of the batch writing ( which blocks for a long time ).